### PR TITLE
[FS] Remove in-progress transactions while reorging

### PR DESCRIPTION
**Pull request by codingupastorm**

### CODE REVIEW

File: src/Stratis.Features.FederatedPeg/Wallet/FederationWalletSyncManager.cs [hunk 1]
```diff
@@ -144,6 +144,7 @@ private async Task OnProcessBlockAsync(Block block, CancellationToken cancellati
 
                     this.logger.LogInformation("Reorg detected, going back from '{0}' to '{1}'.", this.walletTip, fork);
 
+                    this.walletManager.RemoveTransientTransactions();
```
Comment (dangershony) [inline src/Stratis.Features.FederatedPeg/Wallet/FederationWalletSyncManager.cs]: As @quantumagi pointed out won't we need to remove also reorged trx?

**Comment by quantumagi** [comment]
If this happens in the wallet it should happen inside of DeleteBlock for only the relevant transactions.

File: src/Stratis.Features.FederatedPeg/Wallet/FederationWalletSyncManager.cs (hunk 1 above)
Comment (dangershony) [reply src/Stratis.Features.FederatedPeg/Wallet/FederationWalletSyncManager.cs]: I think the Fed wallet is missing a similar functionality as this in the mainchain wallet `RefreshInputKeysLookupLock`

**Comment by quantumagi** [comment]
Also just for reference note that the store also calls `RemoveBlocks`:

image

**Comment by codingupastorm** [comment]
Yo @quantumagi that's not correct. We want to get rid of ALL in-progress transactions during a reorg, the transaction / deposit matching needs to start from scratch.

**Comment by codingupastorm** [comment]
Thanks for the mention of input keys updating. This is useful, I'll add it.

**Comment by quantumagi** [comment]
Please include a test case that builds two partial transfers with UTXO's spanning two consecutive blocks such that the first transactions UTXO come from block A and the second transactions UTXO's come from block B. Set it up so that the first transfer is based on the source chain block X and the second on source chain block Y. After rewinding block B the NextMatureDepositHeight should point at block Y. Now redo the transfer associated with block Y and check that the wallet contains no duplicates. Verify that the changes in this PR improve the outcome.

**Comment by zeptin** [comment]
Our integration tests for the federation are currently skipped, even for less complex scenarios. So while a test for that case is desirable, I doubt it would be more stable than the tests we are currently skipping.

**Comment by quantumagi** [comment]
@zeptin, its not an integration tests and is comparable in complexity and function to the other tests in the `CrossChainTransferStoreTests` class that it will be joining. In fact those can serve as a good starting point.

**Comment by dangershony** [comment]
>We want to get rid of ALL in-progress transactions during a reorg,

If you mean for deposits to the target chain yes we want to get rid of all of them down to the fork point, however thous should be done in the method `RefreshInputKeysLookupLock` inside `wallet.Remove(blocks)`

**Comment by codingupastorm** [comment]
`RefreshInputKeysLookupLock` doesn't actually do that. It just changes the lookups.

I'll add it in, but we still need the `RemoveTransientTransactions()`.

I'd really like to add tests but this is something that occurs sporadically, I don't know that it will be easily possible.

**Comment by quantumagi** [comment]
@codingupastorm , I've added the test case to PR 3560. Please merge it into this PR.

**Comment by quantumagi** [comment]
I've moved the important code from this PR across to #3560.

**Comment by codingupastorm** [comment]
Assuming that #3560 fixes this so closing.

**Pull request closed without merging**